Modules/_zstd/compressor.c

```diff
@@ -0,0 +1,670 @@
+                zd = (ZstdDict*)PyTuple_GET_ITEM(dict, 0);
+                goto load;
+            }
+        }
+    }
+
+    /* Wrong type */
+    PyErr_SetString(PyExc_TypeError,
+                    "zstd_dict argument should be ZstdDict object.");
+    return -1;
+
+load:
+    if (type == DICT_TYPE_DIGESTED) {
+        /* Get ZSTD_CDict */
+        ZSTD_CDict *c_dict = _get_CDict(zd, self->compression_level);
+        if (c_dict == NULL) {
+            return -1;
+        }
+        /* Reference a prepared dictionary.
+           It overrides some compression context's parameters. */
+        Py_BEGIN_CRITICAL_SECTION(self);
+        zstd_ret = ZSTD_CCtx_refCDict(self->cctx, c_dict);
+        Py_END_CRITICAL_SECTION();
+    } else if (type == DICT_TYPE_UNDIGESTED) {
+        /* Load a dictionary.
+           It doesn't override compression context's parameters. */
+        Py_BEGIN_CRITICAL_SECTION2(self, zd);
+        zstd_ret = ZSTD_CCtx_loadDictionary(
+                            self->cctx,
+                            PyBytes_AS_STRING(zd->dict_content),
+                            Py_SIZE(zd->dict_content));
+        Py_END_CRITICAL_SECTION2();
+    } else if (type == DICT_TYPE_PREFIX) {
+        /* Load a prefix */
+        Py_BEGIN_CRITICAL_SECTION2(self, zd);
+        zstd_ret = ZSTD_CCtx_refPrefix(
+                            self->cctx,
+                            PyBytes_AS_STRING(zd->dict_content),
+                            Py_SIZE(zd->dict_content));
+        Py_END_CRITICAL_SECTION2();
+    } else {
+        /* Impossible code path */
+        PyErr_SetString(PyExc_SystemError,
+                        "load_c_dict() impossible code path");
+        return -1;
+    }
+
+    /* Check error */
+    if (ZSTD_isError(zstd_ret)) {
+        set_zstd_error(_module_state, ERR_LOAD_C_DICT, zstd_ret);
+        return -1;
+    }
+    return 0;
+}
+
+#define clinic_state() (get_zstd_state(type))
+#include "clinic/compressor.c.h"
+#undef clinic_state
+
+static PyObject *
+_zstd_ZstdCompressor_new(PyTypeObject *type, PyObject *Py_UNUSED(args), PyObject *Py_UNUSED(kwargs))
+{
+    ZstdCompressor *self;
+    self = (ZstdCompressor*)type->tp_alloc(type, 0);
+    if (self == NULL) {
+        goto error;
+    }
+
+    assert(self->dict == NULL);
+    assert(self->use_multithread == 0);
+    assert(self->compression_level == 0);
+    assert(self->inited == 0);
+
+    /* Compression context */
+    self->cctx = ZSTD_createCCtx();
+    if (self->cctx == NULL) {
+        _zstd_state* const _module_state = PyType_GetModuleState(Py_TYPE(self));
+        if (_module_state != NULL) {
+            PyErr_SetString(_module_state->ZstdError,
+                            "Unable to create ZSTD_CCtx instance.");
+        }
+        goto error;
+    }
+
+    /* Last mode */
+    self->last_mode = ZSTD_e_end;
+
+    return (PyObject*)self;
+
+error:
+    Py_XDECREF(self);
+    return NULL;
+}
+
+static void
+ZstdCompressor_dealloc(ZstdCompressor *self)
```

Comment:
Methods for PyObjects must always take a `PyObject *` as `self` and make the cast inside otherwise you'll get an "incorrect function signature" as it's an undefined behavior if we call this function with a `PyObject*` at runtime.